## 예외 처리
* 자바에서는 실행시 발생할 수 있는 프로그램 오류를 `error`와 `exception`으로 분류합니다.
  * error는 발생하면 복구 할 수 없는 심각한 오류를 말합니다.
  * exception은 발생하더라도 수습될 수 있는 것을 말합니다. 
* RuntimeException 클래스들은 주로 프로그래머의 실수로 발생하는 예외입니다.

### printStackTrace()와 getMessage()
* printStackTrace()를 호출하면 예외 발생 당시 호출 스택에 있었던 메서드의 정보오 예외 메시지 
* getMessage()를 호출하면 Exception 클래스 인스턴스에 저장된 메시지를 얻을 수 있음

### 예외 발생 시키기
```java
throw new Exception("예외 발생");
```

### 메서드에 예외 선언하기
* 예외를 처리하는 방법은 try-catch문을 사용하는 방법도 있지만 메서드에 예외를 선언 할 수도 있습니다.
* 메서드 선언부에 throws 를 사용해서 예외를 선언합니다.
* 이렇게 메서드에 예외를 선언하면 메서드 사용시 예외가 발생할 수 있다고 알릴수 있습니다.
* 메서드를 사용하는 쪽에서 예외에 대한 처리를 맡길 수 있습니다.
```java
void method() throws Exception {
    // 메서드
}
```

### 자동 자원 반환 - try-with-resouces 문
* JDK 1.7 부터 추가되었음 
* 주로 입출력 관련 작업을 할 떄 유용함 
```java
// 기존 try-catch시 이중으로 try-catch문을 잡아야 함
try{
  fis = new FileInputStream("score.dat");
  dis = new DataInputStream(fis);
  //...
} catch( IOException ie){
  ie.printStackTrace();
} finally{
  try{
    if(dis != null)
      dis.close();
  } catch(IOException ie){
    ie.printStackTrace();
  }
}

// 위의 코드를 try-with-resource로 변경하면 
// try ()안에서 객체를 생성하는 문장을 넣으면 close를 따로 하지 않아도 자동으로 호출됨
// 그다음에 catch블록이나 finally블록이 수행됨
try( FileInputStream fis = new FileInputStream("score.dat");
 DataInputStream dis = new DataInputStream(fis) ) {
   while(true){
     score = dis.readInt();
     System.outprint(score);
     sum += score;
   }
} catch (EOFException e){
  System.out.println("점수의 총합은 " + sum + "입니다.");
} catch (IOException ie){
  ie.printStackTrace();
}

// 위의 try-with-resouce에서 자동으로 close를 호출 할 수 있으려면
// 아래의 인터페이스가 구현되어야 합니다.
public interface AutoCloseble{
  void close() throws Exception;
}
```

## 사용자 정의 예외 만들기
* 보통은 Exception이나 RuntimeException 클래스로부터 상속을 통해서 사용자 정의 클래스를 구현할 수 있음
```java
// 현재는 주로 RuntimeException을 상속받아서 작성한다. 
class MyException extends Exception {
  private final int ERR_CODE;   // 생성자로 초기화

  MyException(String msg, int errCode){
    super(msg);
    ERR_CODE = errCodel
  }

  MyException(String msg){
    this(msg, 100);
  }

  public int getErrCode(){  // 에러 코드를 얻을 수 있는 함수 
    return ERR_CODE;    
  }
}
```

### 연결된 예외
* 한 예외가 다른 예외를 발생시킬 수 있음
```java
try{
  startInstall();
  copyFiles();
} catch(SpaceException e){
  IntallException ie = new InstallException("설치중 예외발생")
  ie.initCause(e);  // SpaceException를 원인 예외로 지정\
  throw ie;
}
//...

// 연결된 예외가 사용되는 될 수 있는 곳은
// RuntimeException으로 예외를 변경하면 checked 예외를 unchecked 예외로 변경 할 수 있음
throw new MemoryException("메모리가 부족합니다."); // 예외 처리 필요

throw new RuntimeException(new MemoryException("메모리가 부족합니다."));
```